e2e/api/api_test.go

```diff
@@ -332,21 +389,113 @@ var _ = Describe("API", func() {
 
 		Describe("UpdateServer", func() {
 
-			It("should return codes.NotFound if doesn't exist", func() {
+			// test all invalid / missing fields (missing key, missing service, missing server)
+			DescribeTable("should return codes.NotFound if doesn't exist", func(update *types.RealServer) {
 				server := &types.RealServer{
-					ServiceID: "service1",
+					ServiceID: service.Id,
 					Key:       validServerKey,
 					Config:    validServerConfig,
 				}
-				_, err := client.UpdateServer(ctx, server)
+				_, err := client.CreateServer(ctx, server)
+				Expect(err).ToNot(HaveOccurred())
+
+				_, err = client.UpdateServer(ctx, update)
 				status, ok := status.FromError(err)
 
 				Expect(ok).To(BeTrue(), "got grpc status error")
 				if ok {
 					Expect(status.Code()).To(Equal(codes.NotFound),
 						"expected NotFound, but got %v", err)
 				}
-			})
+			},
+				Entry("no existing server", &types.RealServer{
+					ServiceID: service.Id,
+					Key: &types.RealServer_Key{
+						Ip:   "172.16.20.5",
+						Port: 7474,
+					},
+					Config: &types.RealServer_Config{Forward: types.ForwardMethod_MASQ},
+				}),
+				Entry("no existing service", &types.RealServer{
+					ServiceID: "nonexistent_service",
+					Key:       validServerKey,
+					Config:    &types.RealServer_Config{Forward: types.ForwardMethod_MASQ},
+				}),
+				Entry("missing key", &types.RealServer{
+					ServiceID: service.Id,
+					Config:    &types.RealServer_Config{Forward: types.ForwardMethod_MASQ},
+				}),
+			)
```

Comment:
Could the config be missing? or any of its individual fields?
What about missing fields for the health check?

Reply:
The config has nothing to do with a NotFound status. Only the key of the server (serviceId + Key) affects that.

Reply:
Missing fields for health check is already verified in createServer and in the next test "should be able to update individual fields".